tools/bwa.py

```diff
@@ -47,7 +48,155 @@ def index(self, inFasta, algorithm=None):
+
+            TODO: With the addition of a third aligner to viral-ngs, the functionality
+            common to this method and to the comparable method in the Novoalign wrapper should
+            be broken out as an "aligner" superclass, capable of aligning bam or fastq files with an arbitrary
+            aligner, while preserving read groups. 
+        """
+        options = options or []
+
+        samtools = tools.samtools.SamtoolsTool()
+
+        # Require exactly one RG
+        rgs = samtools.getReadGroups(inBam)
+        if len(rgs) == 0:
+            raise InvalidBamHeaderError("{} lacks read groups".format(inBam))
+        elif len(rgs) == 1:
+            if not rgid:
+                rgid = list(rgs.keys())[0]
+        elif not rgid:
+            raise InvalidBamHeaderError("{} has {} read groups, but we require exactly one".format(inBam, len(rgs)))
+        if rgid not in rgs:
+            raise InvalidBamHeaderError("{} has read groups, but not {}".format(inBam, rgid))
+
+        headerFile = util.file.mkstempfname('.{}.header.txt'.format(rgid))
+        # Strip inBam to just one RG (if necessary)
+        if len(rgs) == 1:
+            one_rg_inBam = inBam
+            tools.samtools.SamtoolsTool().dumpHeader(one_rg_inBam, headerFile)
+        else:
+            # strip inBam to one read group
+            tmp_bam = util.file.mkstempfname('.onebam.bam')
+            samtools.view(['-b', '-r', rgid], inBam, tmp_bam)
+            # special exit if this file is empty
+            if samtools.count(tmp_bam) == 0:
+                return
+            # simplify BAM header otherwise Novoalign gets confused
+            one_rg_inBam = util.file.mkstempfname('.{}.in.bam'.format(rgid))
+            
+            with open(headerFile, 'wt') as outf:
+                for row in samtools.getHeader(inBam):
+                    if len(row) > 0 and row[0] == '@RG':
+                        if rgid != list(x[3:] for x in row if x.startswith('ID:'))[0]:
+                            # skip all read groups that are not rgid
+                            continue
+                    outf.write('\t'.join(row) + '\n')
+            samtools.reheader(tmp_bam, headerFile, one_rg_inBam)
+            os.unlink(tmp_bam)
+
+        # perform actual alignment
+
+        # get the read group line to give to BWA
+        readgroup_line = ""
+        with open(headerFile) as inf:
+            for line in inf:
+                if line.startswith("@RG"):
+                    readgroup_line = line
+
+        assert len(readgroup_line) > 0
+        
+        aln_bam_prefilter = util.file.mkstempfname('.prefiltered.bam')
+        # rather than reheader the alignment bam file later so it has the readgroup information
+        # from the original bam file, we'll pass the RG line to bwa to write out
+        self.mem(one_rg_inBam, refDb, aln_bam_prefilter, opts=options+['-R', readgroup_line.rstrip("\n").rstrip("\r")], min_qual=min_qual, threads=threads)
```

Comment:
I'm assuming our BWA code is emitting `@HD` properly. `@SQ` should not exist upstream (and will be added by BWA) so that shoudl be fine. `@PG` gets added by every thing that touches it along the way (the ailgner, GATK, etc). You can decide whether you want to add a `@PG` in a future iteration of your human depletion process.

So in sum, `@PG` might be the only thing to watch out for in your input, but otherwise, yeah, might as well do what you're doing?